Clean/Table/Basic.lean

```diff
@@ -125,249 +125,249 @@ end TraceOfLength
+end TableContext
 
-instance {W: ℕ+} [Repr F] :
-    ToString (TableConstraintOperation W S F) where
-  toString
-    | Witness offset _ => "(Witness " ++ reprStr offset ++ ")"
-    | GetRow off => "(GetRow " ++ reprStr off ++ ")"
-    | Allocate {ops, ..} => "(Allocate " ++ reprStr ops ++ ")"
-    | Assign v offset => "(Assign " ++ reprStr v ++ " " ++ reprStr offset ++ ")"
+@[reducible, table_norm, table_assignment_norm]
+def TableConstraint (W: ℕ+) (S : Type → Type) (F : Type) [Field F] [ProvableType S] :=
+  StateM (TableContext W S F)
 
-end TableConstraintOperation
+@[table_assignment_norm]
+def assignment_from_circuit {n} (as: CellAssignment W S) : Operations F n → CellAssignment W S
+  | .empty _ => as
+  | .witness ops m _ => (assignment_from_circuit as ops).push_vars_aux m
+  | .assert ops _ => assignment_from_circuit as ops
+  | .lookup ops _ => assignment_from_circuit as ops
+  | .subcircuit ops s => (assignment_from_circuit as ops).push_vars_aux s.local_length
 
-@[reducible, table_norm]
-def TableConstraint (W: ℕ+) (S : Type → Type) (F : Type) [Field F] [NonEmptyProvableType S] :=
-  StateM (TableContext W S F)
+/--
+A `MonadLift` instance from `Circuit` to `TableConstraint` means that we can just use
+all circuit operations inside a table constraint.
+-/
+@[reducible, table_norm, table_assignment_norm]
+instance : MonadLift (Circuit F) (TableConstraint W S F) where
+  monadLift circuit ctx :=
+    let (a, ops) := circuit ctx.circuit
+    (a, {
+      circuit := ops,
+      -- the updated assignment is computed from a fresh starting circuit, independent of the circuit so far.
+      -- (if we would use `ops` instead of `circuit.operations 0`, we would be redoing previous assignments)
+      assignment := assignment_from_circuit ctx.assignment (circuit.operations 0)
+    })
 
 namespace TableConstraint
-def operations {α: Type} {W: ℕ+} (table : TableConstraint W S F α):
-    List (TableConstraintOperation W S F) :=
-  table .empty |>.snd.operations
+@[reducible, table_norm, table_assignment_norm]
+def final_offset (table : TableConstraint W S F α) : ℕ :=
+  table .empty |>.snd.circuit.offset
+
+@[table_norm]
+def operations (table : TableConstraint W S F α) : Operations F table.final_offset :=
+  table .empty |>.snd.circuit.withLength
 
-def assignment {α: Type} {W: ℕ+} (table : TableConstraint W S F α):
-    CellAssignment W S :=
+@[table_assignment_norm]
+def final_assignment (table : TableConstraint W S F α) : CellAssignment W S :=
   table .empty |>.snd.assignment
 
+@[table_assignment_norm]
+def offset_consistent (table : TableConstraint W S F α) : Prop :=
+  table.final_offset = table.final_assignment.offset
+
+-- construct an env by taking the result of the assignment function for input/output cells,
+-- and allowing arbitrary values for aux cells and invalid variables
+def window_env (table : TableConstraint W S F Unit)
+  (window: TraceOfLength F S W) (aux_env : Environment F) : Environment F :=
```

Comment:
the definition of environment is more complicated now, and this is a source of annoyance when unfolding proofs. usually it's necessary to first unfold constraints with an opaque environment, and separately unfold the environment later.

on the positive side, this definition is much closer to an actual implementation. we no longer assume all variables are assigned to an input/output cell (which was unrealistic and would have prevented completeness)